File: sdk/secureai_sdk.py
"""
SecureAI Privacy Shield SDK
Comprehensive Python SDK for integrating privacy protection with AI agents.
"""

import requests
import json
import time
import logging
from typing import Dict, List, Optional, Union, Any
from dataclasses import dataclass
from enum import Enum
import asyncio
import aiohttp
from functools import wraps

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class RedactionLevel(Enum):
    """Redaction levels for different privacy requirements."""
    BASIC = "basic"      # Email, phone, SSN
    STANDARD = "standard"  # Includes addresses, names
    STRICT = "strict"    # All PII types

class ContentType(Enum):
    """Supported content types for redaction."""
    TEXT = "text"
    CODE = "code"
    JSON = "json"
    PDF = "pdf"
    EMAIL = "email"

@dataclass
class RedactionResult:
    """Result of a redaction operation."""
    redacted_content: str
    original_content: str
    redaction_summary: Dict[str, Any]
    processing_time_ms: float
    cached: bool = False
    error: Optional[str] = None

@dataclass
class ProtectionMetrics:
    """Metrics for protection operations."""
    total_requests: int = 0
    successful_redactions: int = 0
    failed_redactions: int = 0
    total_processing_time_ms: float = 0.0
    cache_hits: int = 0
    cache_misses: int = 0

class SecureAIError(Exception):
    """Base exception for SecureAI SDK errors."""
    pass

class SecureAIShield:
    """
    Main SecureAI Privacy Shield client for protecting AI agent data.
    
    This SDK provides comprehensive PII protection for AI agents, including
    real-time detection, automatic redaction, and compliance features.
    """
    
    def __init__(
        self,
        api_key: str,
        endpoint: str = "https://api.secureai.com",
        redaction_level: RedactionLevel = RedactionLevel.STANDARD,
        enable_cache: bool = True,
        cache_ttl: int = 3600,
        timeout: int = 30,
        max_retries: int = 3,
        custom_patterns: Optional[Dict[str, str]] = None
    ):
        """
        Initialize the SecureAI Shield.
        
        Args:
            api_key: Your SecureAI API key
            endpoint: SecureAI API endpoint
            redaction_level: Level of redaction to apply
            enable_cache: Enable caching for better performance
            cache_ttl: Cache time-to-live in seconds
            timeout: Request timeout in seconds
            max_retries: Maximum retry attempts
            custom_patterns: Custom redaction patterns
        """
        self.api_key = api_key
        self.endpoint = endpoint.rstrip('/')
        self.redaction_level = redaction_level
        self.enable_cache = enable_cache
        self.cache_ttl = cache_ttl
        self.timeout = timeout
        self.max_retries = max_retries
        self.custom_patterns = custom_patterns or {}
        
        # Initialize cache
        self._cache = {}
        self._cache_timestamps = {}
        
        # Initialize metrics
        self.metrics = ProtectionMetrics()
        
        # Session for connection pooling
        self.session = requests.Session()
        self.session.headers.update({
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/json',
            'User-Agent': 'SecureAI-SDK/1.0.0'
        })
        
        logger.info(f"SecureAI Shield initialized with endpoint: {endpoint}")
    
    def redact(
        self,
        content: str,
        content_type: ContentType = ContentType.TEXT,
        user_id: Optional[str] = None,
        use_cache: bool = True
    ) -> RedactionResult:
        """
        Redact sensitive information from content.
        
        Args:
            content: Content to redact
            content_type: Type of content being processed
            user_id: User identifier for tracking
            use_cache: Whether to use caching
            
        Returns:
            RedactionResult with redacted content and metadata
        """
        start_time = time.time()
        
        try:
            # Check cache first
            if use_cache and self.enable_cache:
                cache_key = self._generate_cache_key(content, content_type)
                cached_result = self._get_from_cache(cache_key)
                if cached_result:
                    self.metrics.cache_hits += 1
                    return cached_result
            
            # Prepare request
            payload = {
                "content": content,
                "content_type": content_type.value,
                "redaction_level": self.redaction_level.value,
                "user_id": user_id or "anonymous",
                "use_cache": use_cache
            }
            
            # Add custom patterns if provided
            if self.custom_patterns:
                payload["custom_patterns"] = self.custom_patterns
            
            # Make API request
            response = self._make_request("/api/redact", payload)
            
            # Process response
            processing_time = (time.time() - start_time) * 1000
            
            result = RedactionResult(
                redacted_content=response.get("redacted_content", content),
                original_content=content,
                redaction_summary=response.get("redaction_summary", {}),
                processing_time_ms=processing_time,
                cached=response.get("cached", False)
            )
            
            # Cache result
            if use_cache and self.enable_cache:
                cache_key = self._generate_cache_key(content, content_type)
                self._add_to_cache(cache_key, result)
            
            # Update metrics
            self.metrics.total_requests += 1
            self.metrics.successful_redactions += 1
            self.metrics.total_processing_time_ms += processing_time
            
            return result
            
        except Exception as e:
            # Update metrics
            self.metrics.total_requests += 1
            self.metrics.failed_redactions += 1
            
            logger.error(f"Redaction failed: {e}")
            return RedactionResult(
                redacted_content=content,  # Return original content on error
                original_content=content,
                redaction_summary={},
                processing_time_ms=(time.time() - start_time) * 1000,
                error=str(e)
            )
    
    async def redact_async(
        self,
        content: str,
        content_type: ContentType = ContentType.TEXT,
        user_id: Optional[str] = None,
        use_cache: bool = True
    ) -> RedactionResult:
        """
        Asynchronously redact sensitive information from content.
        
        Args:
            content: Content to redact
            content_type: Type of content being processed
            user_id: User identifier for tracking
            use_cache: Whether to use caching
            
        Returns:
            RedactionResult with redacted content and metadata
        """
        start_time = time.time()
        
        try:
            # Check cache first
            if use_cache and self.enable_cache:
                cache_key = self._generate_cache_key(content, content_type)
                cached_result = self._get_from_cache(cache_key)
                if cached_result:
                    self.metrics.cache_hits += 1
                    return cached_result
            
            # Prepare request
            payload = {
                "content": content,
                "content_type": content_type.value,
                "redaction_level": self.redaction_level.value,
                "user_id": user_id or "anonymous",
                "use_cache": use_cache
            }
            
            # Add custom patterns if provided
            if self.custom_patterns:
                payload["custom_patterns"] = self.custom_patterns
            
            # Make async API request
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    f"{self.endpoint}/api/redact",
                    json=payload,
                    headers={
                        'Authorization': f'Bearer {self.api_key}',
                        'Content-Type': 'application/json'
                    },
                    timeout=aiohttp.ClientTimeout(total=self.timeout)
                ) as response:
                    response_data = await response.json()
            
            # Process response
            processing_time = (time.time() - start_time) * 1000
            
            result = RedactionResult(
                redacted_content=response_data.get("redacted_content", content),
                original_content=content,
                redaction_summary=response_data.get("redaction_summary", {}),
                processing_time_ms=processing_time,
                cached=response_data.get("cached", False)
            )
            
            # Cache result
            if use_cache and self.enable_cache:
                cache_key = self._generate_cache_key(content, content_type)
                self._add_to_cache(cache_key, result)
            
            # Update metrics
            self.metrics.total_requests += 1
            self.metrics.successful_redactions += 1
            self.metrics.total_processing_time_ms += processing_time
            
            return result
            
        except Exception as e:
            # Update metrics
            self.metrics.total_requests += 1
            self.metrics.failed_redactions += 1
            
            logger.error(f"Async redaction failed: {e}")
            return RedactionResult(
                redacted_content=content,
                original_content=content,
                redaction_summary={},
                processing_time_ms=(time.time() - start_time) * 1000,
                error=str(e)
            )
    
    def redact_batch(
        self,
        contents: List[str],
        content_type: ContentType = ContentType.TEXT,
        user_id: Optional[str] = None,
        use_cache: bool = True
    ) -> List[RedactionResult]:
        """
        Redact multiple content items efficiently.
        
        Args:
            contents: List of content to redact
            content_type: Type of content being processed
            user_id: User identifier for tracking
            use_cache: Whether to use caching
            
        Returns:
            List of RedactionResult objects
        """
        results = []
        
        for content in contents:
            result = self.redact(content, content_type, user_id, use_cache)
            results.append(result)
        
        return results
    
    async def redact_batch_async(
        self,
        contents: List[str],
        content_type: ContentType = ContentType.TEXT,
        user_id: Optional[str] = None,
        use_cache: bool = True
    ) -> List[RedactionResult]:
        """
        Asynchronously redact multiple content items efficiently.
        
        Args:
            contents: List of content to redact
            content_type: Type of content being processed
            user_id: User identifier for tracking
            use_cache: Whether to use caching
            
        Returns:
            List of RedactionResult objects
        """
        tasks = [
            self.redact_async(content, content_type, user_id, use_cache)
            for content in contents
        ]
        
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        # Handle exceptions
        processed_results = []
        for result in results:
            if isinstance(result, Exception):
                logger.error(f"Batch redaction failed: {result}")
                processed_results.append(RedactionResult(
                    redacted_content="",
                    original_content="",
                    redaction_summary={},
                    processing_time_ms=0,
                    error=str(result)
                ))
            else:
                processed_results.append(result)
        
        return processed_results
    
    def contains_pii(self, content: str) -> bool:
        """
        Check if content contains PII without redacting.
        
        Args:
            content: Content to check
            
        Returns:
            True if PII is detected, False otherwise
        """
        try:
            payload = {
                "content": content,
                "content_type": ContentType.TEXT.value,
                "check_only": True
            }
            
            response = self._make_request("/api/check", payload)
            return response.get("contains_pii", False)
            
        except Exception as e:
            logger.error(f"PII check failed: {e}")
            return True  # Assume PII present on error for safety
    
    def get_redaction_count(self, content: str) -> int:
        """
        Get the number of redactions performed on content.
        
        Args:
            content: Content to analyze
            
        Returns:
            Number of redactions performed
        """
        try:
            payload = {
                "content": content,
                "content_type": ContentType.TEXT.value,
                "analyze_only": True
            }
            
            response = self._make_request("/api/analyze", payload)
            return response.get("redaction_count", 0)
            
        except Exception as e:
            logger.error(f"Redaction count analysis failed: {e}")
            return 0
    
    def get_redaction_summary(self) -> Dict[str, Any]:
        """
        Get a summary of redaction operations.
        
        Returns:
            Dictionary with redaction summary
        """
        return {
            "total_requests": self.metrics.total_requests,
            "successful_redactions": self.metrics.successful_redactions,
            "failed_redactions": self.metrics.failed_redactions,
            "average_processing_time_ms": (
                self.metrics.total_processing_time_ms / max(self.metrics.total_requests, 1)
            ),
            "cache_hit_rate": (
                self.metrics.cache_hits / max(self.metrics.total_requests, 1)
            ),
            "redaction_level": self.redaction_level.value
        }
    
    def health_check(self) -> bool:
        """
        Check if the SecureAI service is healthy.
        
        Returns:
            True if service is healthy, False otherwise
        """
        try:
            response = self._make_request("/health", {})
            return response.get("status") == "healthy"
        except Exception as e:
            logger.error(f"Health check failed: {e}")
            return False
    
    def _make_request(self, endpoint: str, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Make HTTP request to SecureAI API."""
        url = f"{self.endpoint}{endpoint}"
        
        for attempt in range(self.max_retries):
            try:
                response = self.session.post(
                    url,
                    json=payload,
                    timeout=self.timeout
                )
                response.raise_for_status()
                return response.json()
                
            except requests.exceptions.RequestException as e:
                if attempt == self.max_retries - 1:
                    raise SecureAIError(f"API request failed: {e}")
                
                logger.warning(f"Request failed, retrying ({attempt + 1}/{self.max_retries}): {e}")
                time.sleep(2 ** attempt)  # Exponential backoff
    
    def _generate_cache_key(self, content: str, content_type: ContentType) -> str:
        """Generate cache key for content."""
        import hashlib
        key_data = f"{content}:{content_type.value}:{self.redaction_level.value}"
        return hashlib.md5(key_data.encode()).hexdigest()
    
    def _get_from_cache(self, cache_key: str) -> Optional[RedactionResult]:
        """Get result from cache if valid."""
        if cache_key in self._cache:
            timestamp = self._cache_timestamps.get(cache_key, 0)
            if time.time() - timestamp < self.cache_ttl:
                self.metrics.cache_hits += 1
                return self._cache[cache_key]
            else:
                # Remove expired cache entry
                del self._cache[cache_key]
                del self._cache_timestamps[cache_key]
        
        self.metrics.cache_misses += 1
        return None
    
    def _add_to_cache(self, cache_key: str, result: RedactionResult):
        """Add result to cache."""
        self._cache[cache_key] = result
        self._cache_timestamps[cache_key] = time.time()
        
        # Clean up old cache entries
        current_time = time.time()
        expired_keys = [
            key for key, timestamp in self._cache_timestamps.items()
            if current_time - timestamp > self.cache_ttl
        ]
        
        for key in expired_keys:
            del self._cache[key]
            del self._cache_timestamps[key]
    
    def clear_cache(self):
        """Clear the cache."""
        self._cache.clear()
        self._cache_timestamps.clear()
        logger.info("Cache cleared")
    
    def __enter__(self):
        """Context manager entry."""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.session.close()


# Decorator for automatic protection
def protect_ai_agent(shield: SecureAIShield, content_type: ContentType = ContentType.TEXT):
    """
    Decorator to automatically protect AI agent functions.
    
    Args:
        shield: SecureAI Shield instance
        content_type: Type of content being processed
    """
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            # Extract content from arguments
            content = None
            if args:
                content = args[0]
            elif 'content' in kwargs:
                content = kwargs['content']
            elif 'message' in kwargs:
                content = kwargs['message']
            elif 'input' in kwargs:
                content = kwargs['input']
            
            if content and isinstance(content, str):
                # Protect input
                protected_content = shield.redact(content, content_type)
                
                # Update arguments with protected content
                if args:
                    args = (protected_content.redacted_content,) + args[1:]
                else:
                    kwargs['content'] = protected_content.redacted_content
                
                # Call original function
                result = func(*args, **kwargs)
                
                # Protect output if it's a string
                if isinstance(result, str):
                    protected_result = shield.redact(result, content_type)
                    return protected_result.redacted_content
                
                return result
            
            return func(*args, **kwargs)
        
        return wrapper
    return decorator


# Utility functions
def create_shield(
    api_key: str,
    endpoint: Optional[str] = None,
    redaction_level: RedactionLevel = RedactionLevel.STANDARD,
    **kwargs
) -> SecureAIShield:
    """
    Create a SecureAI Shield instance with default settings.
    
    Args:
        api_key: Your SecureAI API key
        endpoint: SecureAI API endpoint (optional)
        redaction_level: Level of redaction to apply
        **kwargs: Additional configuration options
        
    Returns:
        Configured SecureAIShield instance
    """
    return SecureAIShield(
        api_key=api_key,
        endpoint=endpoint or "https://api.secureai.com",
        redaction_level=redaction_level,
        **kwargs
    )


def quick_redact(
    content: str,
    api_key: str,
    redaction_level: RedactionLevel = RedactionLevel.STANDARD
) -> str:
    """
    Quick redaction function for simple use cases.
    
    Args:
        content: Content to redact
        api_key: Your SecureAI API key
        redaction_level: Level of redaction to apply
        
    Returns:
        Redacted content
    """
    with create_shield(api_key, redaction_level=redaction_level) as shield:
        result = shield.redact(content)
        return result.redacted_content


# Example usage
if __name__ == "__main__":
    # Example 1: Basic usage
    shield = SecureAIShield(api_key="your_api_key_here")
    
    content = "My email is john.doe@company.com and my phone is 555-123-4567"
    result = shield.redact(content)
    print(f"Original: {content}")
    print(f"Redacted: {result.redacted_content}")
    print(f"Summary: {result.redaction_summary}")
    
    # Example 2: With decorator
    @protect_ai_agent(shield)
    def ai_chat(message: str) -> str:
        # Simulate AI processing
        return f"AI response to: {message}"
    
    response = ai_chat("My SSN is 123-45-6789")
    print(f"Protected AI response: {response}")
    
    # Example 3: Quick redaction
    quick_result = quick_redact(
        "Contact me at jane.smith@email.com",
        api_key="your_api_key_here"
    )
    print(f"Quick redaction: {quick_result}") 

File: sdk/secureai-sdk.js
/**
 * SecureAI Privacy Shield SDK
 * Comprehensive JavaScript/Node.js SDK for integrating privacy protection with AI agents.
 */

const https = require('https');
const http = require('http');
const { URL } = require('url');

/**
 * Redaction levels for different privacy requirements
 */
const RedactionLevel = {
    BASIC: 'basic',      // Email, phone, SSN
    STANDARD: 'standard', // Includes addresses, names
    STRICT: 'strict'      // All PII types
};

/**
 * Supported content types for redaction
 */
const ContentType = {
    TEXT: 'text',
    CODE: 'code',
    JSON: 'json',
    PDF: 'pdf',
    EMAIL: 'email'
};

/**
 * Result of a redaction operation
 */
class RedactionResult {
    constructor(redactedContent, originalContent, redactionSummary, processingTimeMs, cached = false, error = null) {
        this.redactedContent = redactedContent;
        this.originalContent = originalContent;
        this.redactionSummary = redactionSummary;
        this.processingTimeMs = processingTimeMs;
        this.cached = cached;
        this.error = error;
    }
}

/**
 * Metrics for protection operations
 */
class ProtectionMetrics {
    constructor() {
        this.totalRequests = 0;
        this.successfulRedactions = 0;
        this.failedRedactions = 0;
        this.totalProcessingTimeMs = 0;
        this.cacheHits = 0;
        this.cacheMisses = 0;
    }

    getAverageProcessingTime() {
        return this.totalRequests > 0 ? this.totalProcessingTimeMs / this.totalRequests : 0;
    }

    getCacheHitRate() {
        return this.totalRequests > 0 ? this.cacheHits / this.totalRequests : 0;
    }
}

/**
 * Base exception for SecureAI SDK errors
 */
class SecureAIError extends Error {
    constructor(message) {
        super(message);
        this.name = 'SecureAIError';
    }
}

/**
 * Main SecureAI Privacy Shield client for protecting AI agent data
 */
class SecureAIShield {
    /**
     * Initialize the SecureAI Shield
     * @param {Object} config - Configuration object
     * @param {string} config.apiKey - Your SecureAI API key
     * @param {string} config.endpoint - SecureAI API endpoint (default: https://api.secureai.com)
     * @param {string} config.redactionLevel - Level of redaction to apply (default: STANDARD)
     * @param {boolean} config.enableCache - Enable caching for better performance (default: true)
     * @param {number} config.cacheTtl - Cache time-to-live in seconds (default: 3600)
     * @param {number} config.timeout - Request timeout in seconds (default: 30)
     * @param {number} config.maxRetries - Maximum retry attempts (default: 3)
     * @param {Object} config.customPatterns - Custom redaction patterns
     */
    constructor(config) {
        this.apiKey = config.apiKey;
        this.endpoint = config.endpoint || 'https://api.secureai.com';
        this.redactionLevel = config.redactionLevel || RedactionLevel.STANDARD;
        this.enableCache = config.enableCache !== false;
        this.cacheTtl = config.cacheTtl || 3600;
        this.timeout = config.timeout || 30000;
        this.maxRetries = config.maxRetries || 3;
        this.customPatterns = config.customPatterns || {};

        // Initialize cache
        this._cache = new Map();
        this._cacheTimestamps = new Map();

        // Initialize metrics
        this.metrics = new ProtectionMetrics();

        // Parse endpoint URL
        this.url = new URL(this.endpoint);

        console.log(`SecureAI Shield initialized with endpoint: ${this.endpoint}`);
    }

    /**
     * Redact sensitive information from content
     * @param {string} content - Content to redact
     * @param {string} contentType - Type of content being processed (default: TEXT)
     * @param {string} userId - User identifier for tracking
     * @param {boolean} useCache - Whether to use caching (default: true)
     * @returns {Promise<RedactionResult>} RedactionResult with redacted content and metadata
     */
    async redact(content, contentType = ContentType.TEXT, userId = null, useCache = true) {
        const startTime = Date.now();

        try {
            // Check cache first
            if (useCache && this.enableCache) {
                const cacheKey = this._generateCacheKey(content, contentType);
                const cachedResult = this._getFromCache(cacheKey);
                if (cachedResult) {
                    this.metrics.cacheHits++;
                    return cachedResult;
                }
            }

            // Prepare request payload
            const payload = {
                content: content,
                content_type: contentType,
                redaction_level: this.redactionLevel,
                user_id: userId || 'anonymous',
                use_cache: useCache
            };

            // Add custom patterns if provided
            if (Object.keys(this.customPatterns).length > 0) {
                payload.custom_patterns = this.customPatterns;
            }

            // Make API request
            const response = await this._makeRequest('/api/redact', payload);

            // Process response
            const processingTime = Date.now() - startTime;

            const result = new RedactionResult(
                response.redacted_content || content,
                content,
                response.redaction_summary || {},
                processingTime,
                response.cached || false
            );

            // Cache result
            if (useCache && this.enableCache) {
                const cacheKey = this._generateCacheKey(content, contentType);
                this._addToCache(cacheKey, result);
            }

            // Update metrics
            this.metrics.totalRequests++;
            this.metrics.successfulRedactions++;
            this.metrics.totalProcessingTimeMs += processingTime;

            return result;

        } catch (error) {
            // Update metrics
            this.metrics.totalRequests++;
            this.metrics.failedRedactions++;

            console.error(`Redaction failed: ${error.message}`);
            return new RedactionResult(
                content, // Return original content on error
                content,
                {},
                Date.now() - startTime,
                false,
                error.message
            );
        }
    }

    /**
     * Redact multiple content items efficiently
     * @param {string[]} contents - List of content to redact
     * @param {string} contentType - Type of content being processed (default: TEXT)
     * @param {string} userId - User identifier for tracking
     * @param {boolean} useCache - Whether to use caching (default: true)
     * @returns {Promise<RedactionResult[]>} List of RedactionResult objects
     */
    async redactBatch(contents, contentType = ContentType.TEXT, userId = null, useCache = true) {
        const results = [];

        for (const content of contents) {
            const result = await this.redact(content, contentType, userId, useCache);
            results.push(result);
        }

        return results;
    }

    /**
     * Check if content contains PII without redacting
     * @param {string} content - Content to check
     * @returns {Promise<boolean>} True if PII is detected, False otherwise
     */
    async containsPII(content) {
        try {
            const payload = {
                content: content,
                content_type: ContentType.TEXT,
                check_only: true
            };

            const response = await this._makeRequest('/api/check', payload);
            return response.contains_pii || false;

        } catch (error) {
            console.error(`PII check failed: ${error.message}`);
            return true; // Assume PII present on error for safety
        }
    }

    /**
     * Get the number of redactions performed on content
     * @param {string} content - Content to analyze
     * @returns {Promise<number>} Number of redactions performed
     */
    async getRedactionCount(content) {
        try {
            const payload = {
                content: content,
                content_type: ContentType.TEXT,
                analyze_only: true
            };

            const response = await this._makeRequest('/api/analyze', payload);
            return response.redaction_count || 0;

        } catch (error) {
            console.error(`Redaction count analysis failed: ${error.message}`);
            return 0;
        }
    }

    /**
     * Get a summary of redaction operations
     * @returns {Object} Dictionary with redaction summary
     */
    getRedactionSummary() {
        return {
            total_requests: this.metrics.totalRequests,
            successful_redactions: this.metrics.successfulRedactions,
            failed_redactions: this.metrics.failedRedactions,
            average_processing_time_ms: this.metrics.getAverageProcessingTime(),
            cache_hit_rate: this.metrics.getCacheHitRate(),
            redaction_level: this.redactionLevel
        };
    }

    /**
     * Check if the SecureAI service is healthy
     * @returns {Promise<boolean>} True if service is healthy, False otherwise
     */
    async healthCheck() {
        try {
            const response = await this._makeRequest('/health', {});
            return response.status === 'healthy';
        } catch (error) {
            console.error(`Health check failed: ${error.message}`);
            return false;
        }
    }

    /**
     * Make HTTP request to SecureAI API
     * @param {string} endpoint - API endpoint
     * @param {Object} payload - Request payload
     * @returns {Promise<Object>} API response
     */
    async _makeRequest(endpoint, payload) {
        const url = `${this.endpoint}${endpoint}`;
        const urlObj = new URL(url);

        for (let attempt = 0; attempt < this.maxRetries; attempt++) {
            try {
                const result = await this._makeHttpRequest(urlObj, payload);
                return result;
            } catch (error) {
                if (attempt === this.maxRetries - 1) {
                    throw new SecureAIError(`API request failed: ${error.message}`);
                }

                console.warn(`Request failed, retrying (${attempt + 1}/${this.maxRetries}): ${error.message}`);
                await this._sleep(Math.pow(2, attempt) * 1000); // Exponential backoff
            }
        }
    }

    /**
     * Make HTTP request with proper error handling
     * @param {URL} urlObj - Parsed URL object
     * @param {Object} payload - Request payload
     * @returns {Promise<Object>} API response
     */
    _makeHttpRequest(urlObj, payload) {
        return new Promise((resolve, reject) => {
            const postData = JSON.stringify(payload);

            const options = {
                hostname: urlObj.hostname,
                port: urlObj.port || (urlObj.protocol === 'https:' ? 443 : 80),
                path: urlObj.pathname + urlObj.search,
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'Content-Type': 'application/json',
                    'User-Agent': 'SecureAI-SDK/1.0.0',
                    'Content-Length': Buffer.byteLength(postData)
                },
                timeout: this.timeout
            };

            const client = urlObj.protocol === 'https:' ? https : http;

            const req = client.request(options, (res) => {
                let data = '';

                res.on('data', (chunk) => {
                    data += chunk;
                });

                res.on('end', () => {
                    try {
                        if (res.statusCode >= 200 && res.statusCode < 300) {
                            const response = JSON.parse(data);
                            resolve(response);
                        } else {
                            reject(new Error(`HTTP ${res.statusCode}: ${data}`));
                        }
                    } catch (error) {
                        reject(new Error(`Failed to parse response: ${error.message}`));
                    }
                });
            });

            req.on('error', (error) => {
                reject(error);
            });

            req.on('timeout', () => {
                req.destroy();
                reject(new Error('Request timeout'));
            });

            req.write(postData);
            req.end();
        });
    }

    /**
     * Generate cache key for content
     * @param {string} content - Content to cache
     * @param {string} contentType - Type of content
     * @returns {string} Cache key
     */
    _generateCacheKey(content, contentType) {
        const crypto = require('crypto');
        const keyData = `${content}:${contentType}:${this.redactionLevel}`;
        return crypto.createHash('md5').update(keyData).digest('hex');
    }

    /**
     * Get result from cache if valid
     * @param {string} cacheKey - Cache key
     * @returns {RedactionResult|null} Cached result or null
     */
    _getFromCache(cacheKey) {
        if (this._cache.has(cacheKey)) {
            const timestamp = this._cacheTimestamps.get(cacheKey) || 0;
            if (Date.now() - timestamp < this.cacheTtl * 1000) {
                this.metrics.cacheHits++;
                return this._cache.get(cacheKey);
            } else {
                // Remove expired cache entry
                this._cache.delete(cacheKey);
                this._cacheTimestamps.delete(cacheKey);
            }
        }

        this.metrics.cacheMisses++;
        return null;
    }

    /**
     * Add result to cache
     * @param {string} cacheKey - Cache key
     * @param {RedactionResult} result - Result to cache
     */
    _addToCache(cacheKey, result) {
        this._cache.set(cacheKey, result);
        this._cacheTimestamps.set(cacheKey, Date.now());

        // Clean up old cache entries
        const currentTime = Date.now();
        for (const [key, timestamp] of this._cacheTimestamps.entries()) {
            if (currentTime - timestamp > this.cacheTtl * 1000) {
                this._cache.delete(key);
                this._cacheTimestamps.delete(key);
            }
        }
    }

    /**
     * Clear the cache
     */
    clearCache() {
        this._cache.clear();
        this._cacheTimestamps.clear();
        console.log('Cache cleared');
    }

    /**
     * Sleep utility function
     * @param {number} ms - Milliseconds to sleep
     * @returns {Promise} Promise that resolves after sleep
     */
    _sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

/**
 * Decorator to automatically protect AI agent functions
 * @param {SecureAIShield} shield - SecureAI Shield instance
 * @param {string} contentType - Type of content being processed
 * @returns {Function} Decorated function
 */
function protectAIAgent(shield, contentType = ContentType.TEXT) {
    return function(target, propertyKey, descriptor) {
        const originalMethod = descriptor.value;

        descriptor.value = async function(...args) {
            // Extract content from arguments
            let content = null;
            if (args.length > 0) {
                content = args[0];
            } else if (this.content) {
                content = this.content;
            } else if (this.message) {
                content = this.message;
            } else if (this.input) {
                content = this.input;
            }

            if (content && typeof content === 'string') {
                // Protect input
                const protectedContent = await shield.redact(content, contentType);

                // Update arguments with protected content
                if (args.length > 0) {
                    args[0] = protectedContent.redactedContent;
                } else {
                    this.content = protectedContent.redactedContent;
                }

                // Call original function
                const result = await originalMethod.apply(this, args);

                // Protect output if it's a string
                if (typeof result === 'string') {
                    const protectedResult = await shield.redact(result, contentType);
                    return protectedResult.redactedContent;
                }

                return result;
            }

            return originalMethod.apply(this, args);
        };

        return descriptor;
    };
}

/**
 * Create a SecureAI Shield instance with default settings
 * @param {string} apiKey - Your SecureAI API key
 * @param {Object} config - Additional configuration options
 * @returns {SecureAIShield} Configured SecureAIShield instance
 */
function createShield(apiKey, config = {}) {
    return new SecureAIShield({
        apiKey,
        endpoint: config.endpoint || 'https://api.secureai.com',
        redactionLevel: config.redactionLevel || RedactionLevel.STANDARD,
        ...config
    });
}

/**
 * Quick redaction function for simple use cases
 * @param {string} content - Content to redact
 * @param {string} apiKey - Your SecureAI API key
 * @param {string} redactionLevel - Level of redaction to apply (default: STANDARD)
 * @returns {Promise<string>} Redacted content
 */
async function quickRedact(content, apiKey, redactionLevel = RedactionLevel.STANDARD) {
    const shield = createShield(apiKey, { redactionLevel });
    const result = await shield.redact(content);
    return result.redactedContent;
}

// Export classes and functions
module.exports = {
    SecureAIShield,
    RedactionResult,
    ProtectionMetrics,
    SecureAIError,
    RedactionLevel,
    ContentType,
    protectAIAgent,
    createShield,
    quickRedact
};

// Example usage
if (require.main === module) {
    (async () => {
        // Example 1: Basic usage
        const shield = new SecureAIShield({ apiKey: 'your_api_key_here' });

        const content = 'My email is john.doe@company.com and my phone is 555-123-4567';
        const result = await shield.redact(content);
        console.log(`Original: ${content}`);
        console.log(`Redacted: ${result.redactedContent}`);
        console.log(`Summary: ${JSON.stringify(result.redactionSummary)}`);

        // Example 2: With decorator
        class AIAgent {
            constructor(shield) {
                this.shield = shield;
            }

            @protectAIAgent(shield)
            async chat(message) {
                // Simulate AI processing
                return `AI response to: ${message}`;
            }
        }

        const aiAgent = new AIAgent(shield);
        const response = await aiAgent.chat('My SSN is 123-45-6789');
        console.log(`Protected AI response: ${response}`);

        // Example 3: Quick redaction
        const quickResult = await quickRedact(
            'Contact me at jane.smith@email.com',
            'your_api_key_here'
        );
        console.log(`Quick redaction: ${quickResult}`);
    })().catch(console.error);
} 